src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1436 @@
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+    N_VLinearSum(ONE, ark_mem->ycur, sm1inv * ark_mem->h, ark_mem->fn,
+                 ark_mem->ycur);
+    N_VLinearSum(ONE, ark_mem->tempv1, bt2 * ark_mem->h, ark_mem->fn,
+                 ark_mem->tempv1);
+  }
+  /* Evaluate the last stage for j = step_mem->reqstages */
+  retval = step_mem->fe(ark_mem->tcur + ark_mem->h, ark_mem->ycur, ark_mem->fn,
+                        ark_mem->user_data);
+  if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+
+  N_VLinearSum(ONE / rs, ark_mem->yn, ONE / (sm1inv * rs), ark_mem->ycur,
+               ark_mem->ycur);
+  N_VLinearSum(ONE, ark_mem->ycur, ark_mem->h / rs, ark_mem->fn, ark_mem->ycur);
+
+  N_VLinearSum(ONE, ark_mem->tempv1, bt3 * ark_mem->h, ark_mem->fn,
+               ark_mem->tempv1);
+
+  step_mem->nfe += step_mem->reqstages - 1;
+
+  /* Compute yerr (if step adaptivity enabled) */
+  if (!ark_mem->fixedstep)
+  {
+    N_VLinearSum(ONE, ark_mem->ycur, -ONE, ark_mem->tempv1, ark_mem->tempv1);
+
+    *dsmPtr = N_VWrmsNorm(ark_mem->tempv1, ark_mem->ewt);
+  }
+  if (*dsmPtr <= ONE || ark_mem->fixedstep)
+  {
+    retval = step_mem->fe(ark_mem->tcur + ark_mem->h, ark_mem->ycur,
+                          ark_mem->fn, ark_mem->user_data);
+    ark_mem->fn_is_current = SUNTRUE;
+    step_mem->nfe++;
+    if (retval != ARK_SUCCESS) { return (ARK_RHSFUNC_FAIL); }
+  }
+
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_TakeStepSSPs3:
+
+  This routine serves the primary purpose of the LSRKStepSSPs3 module:
+  it performs a single SSPs3 step (with embedding).
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is used to gauge convergence
+  of any algebraic solvers within the step.  As this routine
+  involves no algebraic solve, it is set to 0 (success).
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepSSPs3(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+
+  ARKodeLSRKStepMem step_mem;
+
+  /* initialize algebraic solver convergence flag to success,
+     temporal error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  sunrealtype rs  = (sunrealtype)step_mem->reqstages;
+  sunrealtype rn  = SUNRsqrt(rs);
+  sunrealtype rat = ONE / (rs - rn);
+  int in          = (int)rn;
```

Comment:
Does it matter whether `rn` is rounded up vs down vs to the nearest integer in this cast?

Reply:
I wonder if we need a new SUNDIALS function for this, that would use `lround` or `lroundf` (or similar) based on the precision of `sunrealtype`?  I'm concerned that the direct cast here will always round *down* instead of to the nearest integer.